internal/brokerstore/brokerstore.go

```diff
@@ -81,55 +79,54 @@ func NewBrokerStore(variant string, config *messagebroker.BrokerConfig) (IBroker
 	}
 
 	return &BrokerStore{
-		variant: variant,
-		bConfig: config,
+		variant:       variant,
+		bConfig:       config,
+		partitionLock: partitionlocker.NewPartitionLocker(&sync.Mutex{}),
 	}, nil
 }
 
 // GetOrCreateConsumer returns for an existing consumer instance, if available returns that else creates as new instance
-func (b *BrokerStore) GetOrCreateConsumer(ctx context.Context, op messagebroker.ConsumerClientOptions) (messagebroker.Consumer, error) {
-	key := NewKey(b.variant, op.Partition)
-
-	if consumer, exists := b.consumerMap.Load(key.String()); exists {
+func (b *BrokerStore) GetOrCreateConsumer(ctx context.Context, id string, op messagebroker.ConsumerClientOptions) (messagebroker.Consumer, error) {
+	key := NewKey(op.GroupID, id)
+	consumer, ok := b.consumerMap.Load(key.String())
+	if ok {
+		return consumer.(messagebroker.Consumer), nil
+	}
+	b.partitionLock.Lock(key.String())              // lock
+	consumer, ok = b.consumerMap.Load(key.String()) // double-check
+	if ok {
 		return consumer.(messagebroker.Consumer), nil
 	}
-
 	newConsumer, perr := messagebroker.NewConsumerClient(ctx,
 		b.variant,
+		id,
 		b.bConfig,
 		&messagebroker.ConsumerClientOptions{Topic: op.Topic, Subscription: op.Subscription, GroupID: op.GroupID},
 	)
-
 	if perr != nil {
 		return nil, perr
 	}
-
-	b.consumerMap.Store(key, newConsumer)
-
-	return newConsumer, nil
+	consumer, _ = b.consumerMap.LoadOrStore(key.String(), newConsumer)
+	b.partitionLock.Unlock(key.String()) // unlock
+	return consumer.(messagebroker.Consumer), nil
 }
 
 // GetOrCreateProducer returns for an existing producer instance, if available returns that else creates as new instance
 func (b *BrokerStore) GetOrCreateProducer(ctx context.Context, op messagebroker.ProducerClientOptions) (messagebroker.Producer, error) {
-	key := NewKey(b.variant, op.Partition)
-
-	if producer, exists := b.producerMap.Load(key.String()); exists {
-		return producer.(messagebroker.Producer), nil
-	}
+	key := NewKey(b.variant, "")
```

Comment:
we seem to be creating a new producer client everytime, didn't got why, if that is require, do we need to keep a map and store them ?

Reply:
Fixed. Added a comment to review this while doing perf